Added: Configuration Builder

Ant-hem opened the pull request:
## Describe your change
This PR is adding a builder pattern to the configuration of the clients.

## What problem is this fixing?
Mutability of the current configuration.

Ant-hem committed:
Added: Configuration Builder

BenoitPerrot commented on src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -1,44 +1,139 @@
-  ConfigBase(String applicationID, String apiKey) {
-    this(applicationID, apiKey, ForkJoinPool.commonPool());
-  }
+class ConfigBase {
```
Nice to have that now internal!
But why not keeping it abstract?

Ant-hem replied:
My bad :( Fixed with 23c1d91

BenoitPerrot commented on src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -1,44 +1,139 @@
+
+    void setDefaultHosts(List<StatefulHost> defaultHosts) {
+      this.defaultHosts = defaultHosts;
+    }
```
"return getThis();"

Ant-hem replied:
👍 fixed with 23c1d91

BenoitPerrot commented on src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -1,44 +1,139 @@
+      return getThis();
+    }
+
+    public ConfigBase build() {
```
hmmmm wait, this "build()" shouldn't be invoked, right?

Ant-hem replied:
Good catch! My bad again.. fixed with 23c1d91

BenoitPerrot commented on src/main/java/com/algolia/search/AnalyticsConfig.java:
```diff
@@ -1,44 +1,57 @@
-
-    this.setCustomHosts(customHosts);
+  private AnalyticsConfig(Builder builder) {
+    super(builder);
```
The pattern I've seen the most: keeping AnalyticsConfig constructor as before (= the constructor with all parameters) but _private_ - or internal, and have the builder invokes it in its "build" method. This way the coupling between the two is loose (which is desirable from a maintenance/evolutivity point of view)

Ant-hem replied:
Yes! I've seen both patterns, I found the current one easier to implement with the current logic of the hosts. But if we decide to drop the Builder constructor with the hosts. i.e having only one with AppID and APIKey it could be nice to have this pattern. (see above for the constructors) :)

BenoitPerrot commented on src/main/java/com/algolia/search/AnalyticsConfig.java:
```diff
@@ -1,44 +1,57 @@
+
+    /**
+     * Creates a {@link AnalyticsConfig} for the search client with custom {@link StatefulHost}.
+     * Warning: Defaults hosts are not set when setting custom {@link StatefulHost}.
```
(It would be awesome to find an API making what is happening - overriding default hosts - clear)

BenoitPerrot replied:
Wait: why not just have a "hosts" list, defaulting to, well, default hosts, with a "setHosts" method? This would make it very clear that default hosts would get overridden.

Ant-hem replied:
Nice! If we are on the same page, we would like to drop the default builder constructor: 

```java
    SearchConfig config =
        new SearchConfig.Builder("appID", "apiKEY", Collections.singletonList(new StatefulHost("Algolia", EnumSet.of(CallType.READ))))
            .build();
```

to the classic builder pattern. I like it :D WDYT?

```java
    SearchConfig config =
        new SearchConfig.Builder("appID", "apiKEY")
            .setCustomHosts(
                Collections.singletonList(new StatefulHost("Algolia", EnumSet.of(CallType.READ))))
            .build();
```

Ant-hem replied:
Or with `setHosts` instead of `setCustomHosts`

Ant-hem replied:
Proposal: c38029d

BenoitPerrot replied:
Yes, that's it: `setHosts`, and Builders' constructors taking only applicationID and apiKey 👍

Ant-hem replied:
Perfect! Thanks :)

BenoitPerrot commented on src/main/java/com/algolia/search/InsightsConfig.java:
```diff
@@ -6,51 +6,66 @@
-    super(applicationID, apiKey);
-
-    this.setCustomHosts(customHosts);
+  private InsightsConfig(Builder builder) {
```
Same as above: I suggest to keep the a private InsightConfig constructor with all parameters and invoke that in "build()"

BenoitPerrot commented on src/test/java/com/algolia/search/integration/client/TimeoutTest.java:
```diff
@@ -16,10 +16,10 @@
-                new SearchConfig(ALGOLIA_APPLICATION_ID_1, ALGOLIA_API_KEY_1)
-                    .setConnectTimeOut(2)
-                    .setReadTimeOut(2))) {
+            new SearchConfig.Builder(ALGOLIA_APPLICATION_ID_1, ALGOLIA_API_KEY_1)
```
Ah! Way better than the forced-downcasting previous version :)

Ant-hem replied:
Sure :D

BenoitPerrot commented on src/main/java/com/algolia/search/SearchConfig.java:
```diff
@@ -9,54 +9,64 @@
-    super(applicationID, apiKey);
-
-    this.setCustomHosts(customHosts);
+  private SearchConfig(Builder builder) {
```
Same as above: I suggest to keep a private constructor with all parameters, and invoke it in "build()"

Ant-hem committed:
Updated: protection level of ConfigBase

Ant-hem committed:
Removed: customHosts constructor in builders
- Also renamed setCustomHosts to setHosts

BenoitPerrot commented on src/main/java/com/algolia/search/AnalyticsConfig.java:
```diff
@@ -1,44 +1,41 @@
+     * @param apiKey The API Key
+     */
+    public Builder(@Nonnull String applicationID, @Nonnull String apiKey) {
+      super(applicationID, apiKey);
```
I suggest to adjust `ConfigBase`'s constructor to make it receive default hosts (rationale: it is not an optional parameter for ConfigBase):
```
super(applicationID, apiKey, Collections.singletonList(...))
```

BenoitPerrot commented on src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -1,44 +1,136 @@
+    private Integer writeTimeOut;
+    private Integer connectTimeOut;
+    private List<StatefulHost> defaultHosts;
+    private List<StatefulHost> customHosts;
```
I suggest to have just: `private List<StatefulHost> hosts;`. No "default" vs "custom" ones.

BenoitPerrot commented on src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -1,44 +1,136 @@
+     * @param apiKey The API Key: could be Admin API Key or Search API Key
+     * @throws NullPointerException if ApplicationID/ApiKey/Config/Requester is null
+     */
+    public Builder(@Nonnull String applicationID, @Nonnull String apiKey) {
```
Here take "defaultHosts" as a parameter, and use it to set the "hosts" attribute (the names I am using here are chosen on purpose: at construction time is given a list of default hosts)

BenoitPerrot commented on src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -1,44 +1,136 @@
+      return getThis();
+    }
+
+    T setDefaultHosts(List<StatefulHost> defaultHosts) {
```
(just remove this one)

BenoitPerrot commented on src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -1,44 +1,136 @@
+
+    /** Sets a list of specific host to target. Default hosts will be overridden. */
+    public T setHosts(List<StatefulHost> customHosts) {
+      this.customHosts = customHosts;
```
`this.hosts = hosts`

BenoitPerrot commented on src/main/java/com/algolia/search/InsightsConfig.java:
```diff
@@ -1,56 +1,59 @@
+     * @param apiKey The API Key
+     */
+    public Builder(@Nonnull String applicationID, @Nonnull String apiKey, @Nonnull String region) {
+      super(applicationID, apiKey);
```
`super(applicationID, apiKey, Collections.singletonList(...))`

BenoitPerrot reviewed (commented):
I propose to drop `setDefaultHosts`, replacing `defaultHosts` and `customHosts` with just `hosts`

BenoitPerrot commented on src/main/java/com/algolia/search/ConfigBase.java:
```diff
@@ -1,44 +1,136 @@
+    }
+
+    /** To prevent unchecked cast warning. */
+    public abstract T getThis();
```
OK, actually there is a simpler construction. Tell me what you think:

```java
abstract class ConfigBase {
  public abstract static class Builder<T> {
    // [...]
    public abstract T build();

    public Builder<T> setBatchSize(int batchSize) {
      this.batchSize = batchSize;
      return this;
    }
    // [...]
  }
}
```

And other Builders:
```java
public final class InsightsConfig extends ConfigBase {
  // [...]
  public static class Builder extends ConfigBase.Builder<InsightsConfig> {
    @Override
    public InsightsConfig build() {
      return new InsightsConfig(this);
    }
  }
  // [...]
}
```

Ant-hem committed:
Removed: customHosts and defaultHosts in favor of hosts
- Added constructor with hosts in ConfigBase
- Calling super in child classes with buildHosts() method

Ant-hem commented:
@BenoitPerrot proposal with all your reviews 2a3538e 
Looks pretty good now :)

BenoitPerrot commented on src/main/java/com/algolia/search/AnalyticsConfig.java:
```diff
@@ -33,6 +27,14 @@ public Builder getThis() {
+
+    /** Create default hosts for the analytics configuration */
+    private static List<StatefulHost> createHosts(
+        @Nonnull String applicationID, @Nonnull String apiKey) {
```
`applicationID` and `apiKey` are unused, I suggest to do without

BenoitPerrot commented on src/main/java/com/algolia/search/InsightsConfig.java:
```diff
@@ -51,6 +40,19 @@ public Builder getThis() {
+
+    /** Create default hosts for the insights configuration */
+    private static List<StatefulHost> createDefaultHosts(
+        @Nonnull String applicationID, @Nonnull String apiKey, @Nonnull String region) {
```
same here: `applicationID` and `apiKey` are unused, I suggest to do without

BenoitPerrot commented on src/main/java/com/algolia/search/SearchConfig.java:
```diff
@@ -13,7 +13,21 @@
+
+    /** Create default hosts for the search configuration */
+    private static List<StatefulHost> createDefaultHosts(
+        @Nonnull String applicationID, @Nonnull String apiKey) {
```
here too: `applicationID` and `apiKey` are unused, I suggest to do without

Ant-hem replied:
Fixed with eb0e845 :)

BenoitPerrot replied:
Hmmmm are you sure? I am still seeing it in the final diff

Ant-hem replied:
Should be better now.

Ant-hem replied:
`applicationID` is used to create the host but `apiKey` is unused.

BenoitPerrot reviewed (changes requested):
`applicationID` and `apiKey` are unused in createDefaultHosts methods: let's do without (if we need it in the future, we would introduce them then)

Ant-hem commented:
> `applicationID` and `apiKey` are unused in createDefaultHosts methods: let's do without (if we need it in the future, we would introduce them then)

Arg. Copy/paste is always bad. Good catch!

Ant-hem committed:
Removed: unused parameters in Analytics/Insights createDefaultHosts()

Ant-hem committed:
Removed: unused ApiKeys parameter in searchConfig/CreateDefault hosts

BenoitPerrot reviewed (approved):
Looks good!

Ant-hem commented:
Thank you so much for you thorough review on this PR 🎉 
We have a neat configuration builder now.
